Cloudwatch: Add syntax highlighting and autocomplete for "Metric Search"

sarahzinger opened the pull request:
**What this PR does / why we need it**:
In the cloudwatch plugin, we allow users to use [AWS's metric math expressions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html) but do not offer much in the way of autocomplete or syntax highlighting. 

before:
<img width="1144" alt="Screen Shot 2022-01-12 at 9 19 22 PM" src="https://user-images.githubusercontent.com/6620164/149254228-ee07c694-de2e-4ed5-827b-9f7acf395092.png">

after:
<img width="1198" alt="Screen Shot 2022-01-12 at 8 59 45 PM" src="https://user-images.githubusercontent.com/6620164/149254258-378094b9-6870-43b4-b4d1-fda652755041.png">


**Which issue(s) this PR fixes**:

Fixes #43984

**Special notes for your reviewer**:

This is a large pr but I tried to break it into 3 understandable commits, so hopefully it's not too bad! If it is too difficult to look at all at once, let me know and I can close this pr and break this work up into smaller PRs. Hopefully this isn't too bad since most of the syntax highlighting has already been reviewed/approved in a private repo a while back.

I went back and forth about how advanced to go with the autocomplete for metric math. I couldn't really come up with a smarter solution without spending more time (and I've already spent a lot of time on this haha), but I think even this fairly heavy handed solution is pretty useful in comparison to what we have, and hopefully it'll be easier to iterate on and improve.

I didn't add much in the way of tests, mostly just moved around existing tests and updated them as interfaces changed, but if we like this approach I was thinking I would add a few tests to ensure getSuggestions returns what we expect for metric math and maybe also one for getStatementPosition as I could see that getting more complicated as time goes on.

sarahzinger committed:
Create a "monarch" folder with everything you need to do syntax highlighting and autocompletion.

sarahzinger committed:
Use this new monarch folder with existing cloudwatch sql.

sarahzinger committed:
Add metric math syntax highlighting and autocomplete.

andresmgot commented:
I tried to give this a spin but I am getting a lot of errors. When I try to open the metric search l get this:

![Screenshot from 2022-01-13 17-27-31](https://user-images.githubusercontent.com/4025665/149369670-ed99ecee-8267-4892-a1f7-d6496cfae009.png)

Looks like I am missing something, do you know what could be?

andresmgot commented:
Great work! I didn't read all the code but I gave this PR a spin. Noticed a couple of things:

It seems that the syntax highlight is not working for me? (at least the colors are different from the ones you posted)

![Screenshot from 2022-01-13 17-35-55](https://user-images.githubusercontent.com/4025665/149373178-c4cc563e-b60e-4fc9-a159-72713d76a5ec.png)

Also, it seems that regardless of the position of the cursor, I am always suggested to use the functions:

![Peek 2022-01-13 17-39](https://user-images.githubusercontent.com/4025665/149373273-d48e7393-4b68-44ca-9db0-fc6284ec341c.gif)

Is that expected? I guess we can have nested functions but within strings, should we avoid suggestions?

sarahzinger commented:
@andresmgot very strange to me about the colors :( maybe we can both look at this together next week and figure out why we're seeing different behavior. 

Regarding autocomplete: it seems you can have nested functions (and that people do this a lot) so I leaned towards over suggesting functions haha, but I agree there's no need to have them within strings! I'll update so that when you're within a string we stop suggesting functions!

sarahzinger committed:
Make autocomplete "smarter":
- search always inserts a string as first arg
- strings can't contain predefined functions
- operators follow the last closing )

sarahzinger committed:
Add some tests for Metric Math's CompletionItemProvider.

sunker commented on public/app/plugins/datasource/cloudwatch/datasource.ts:
```diff
@@ -59,7 +59,18 @@ import { increasingInterval } from './utils/rxjs/increasingInterval';
+import { getSuggestionKinds as getMetricMathSuggestionKinds } from './metric-math/completion/suggestionKind';
+import { getStatementPosition as getMetricMathStatementPosition } from './metric-math/completion/statementPosition';
+import cloudWatchMetricMathLanguageDefinition from './metric-math/definition';
+import { MetricMathTokenType } from './metric-math/completion/types';
```
Instead of importing all these things (line 63-73), can't they be imported in their respective CompletionItemProvider and then pass it along to the super constructor? Then we only need to instantiate MetricMathCompletionItemProvider and SQLCompletionItemProvider with a reference to _this_  and template here.

sarahzinger replied:
@sunker I realized there's really no reason to pass in these args at all since they are static, so I reworked it a bit, hopefully it makes sense, let me know if you want me to walk through any of the changes!

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/completion/CompletionItemProvider.ts:
```diff
@@ -0,0 +1,84 @@
+        label: v,
+        insertText: v,
+        kind: monaco.languages.CompletionItemKind.Variable,
+        sortText: CompletionItemPriority.MediumHigh,
```
I feel like variables should have lower priority than suggestions for functions, operators and keywords. WDYT?

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/completion/statementPosition.ts:
```diff
@@ -0,0 +1,21 @@
+  if (!currentToken || !currentToken.isString()) {
+    return StatementPosition.PredefinedFunction;
+  }
+
```
Not completely necessary, but would be fairly straightforward to detect second and third argument for search expressions. If so, we could suggest [statistics](https://github.com/grafana/grafana/blob/sarahzinger/metric-math-expressions-revamp/public/app/plugins/datasource/cloudwatch/datasource.ts#L112) and periods. Up to you!

sarahzinger replied:
I was able to suggest statistics, but I wasn't sure how to suggest a period of time, it seems to me like it'll just take any number so I'm not sure we can suggest that, but maybe there's a time related macro or something we should suggest? What do you think?

sunker replied:
It's any 60 seconds that can be multiplied by 60. And for a few high resolution metrics, it can be lower than that. But if we can suggest these number that would be sufficient in most scenarios I think. 
`10, 60, 300, 900, 3600, 21600, 86400`

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/completion/CompletionItemProvider.ts:
```diff
@@ -0,0 +1,84 @@
+            addSuggestion(s, {
+              insertText: `${s} `,
+              command: TRIGGER_SUGGEST,
+            })
```
Do we need the insertText template here? 

Can we give these a separate lower (or higher priority) so that they are separated from functions in the list of suggestions? Also, I think there's a CompletionItemKind called _Keyword_ that wer can use.

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/language.ts:
```diff
@@ -0,0 +1,153 @@
+      [/\)/, { token: 'delimiter.parenthesis', next: '@pop' }], // escape out of string and into braces land
+      [/'/, { token: 'string', next: '@string' }], // go to string land if see start of string
+      [/"/, { token: 'type', next: '@string_double' }], // go to string land if see start of string
+    ],
```
Great work on this!!

sunker reviewed (commented):
Great progress! Just added a few minor comments. 

For me the syntax highlighting is working fine. I think the autocomplete experience for SEARCH expressions could be improved pretty significantly by suggesting stats and period. See comment below. 

Would be great if we could replace the CodeEditor component from Grafana UI with something that is more similar to the code editor field used in the Prometheus data source. I.e something that has a default height of one line but will grow as you add new lines.

sarahzinger committed:
Fixes After CR:
- refactor CompletionItemProvider, so that it only requires args that are dynamic or outside of it's responsibility
- Update and add tests with mocked monaco
- Add more autocomplete suggestions for SEARCH expression functions
- sort keywords and give different priority from function to make more visually distinctive.

sarahzinger commented:
@sunker I spent a little time looking into the code editor that Prometheus has that auto expands and despite it seeming pretty clear-cut, I wasn't able to easily port it over or put that code in CodeEditor in grafana/ui. I'm thinking maybe we can make a ticket for it and come back to it and/or look at it together to see what I'm missing. I wonder if the issue seems to be that we Lazy load this editor and that one is not? Idk.

gabor commented:
@sarahzinger hi, i wrote the auto-expand code, and if i can help in some way, just ping me. the prometheus query editor uses lazy loading inside, but it does not use `<CodeEditor/>` so maybe a different approach will be needed there.

sarahzinger commented:
@gabor cool, I'll take another look and set up some time to look at it together if I can't figure it out! Thanks so much

sarahzinger committed:
Change QueryEditor to auto-resize and look more like the one in Prometheus.

sarahzinger committed:
Add autocomplete for time periods for the third arg of Search.

sarahzinger commented:
@sunker ok added auto resizing and period suggestions if you want to take another look!
 
https://recordit.co/2JpwJ9TrJ6 

I started off adding this to grafana/ui's CodeEditor with a prop like `shouldAutoResize` but I'm somewhat on the fence on whether or not that is helpful or if it just makes it more complicated to use. I realized though that there's no reason we can't just keep it in the datasource for now, it's not a lot of code to copy over if we want it in a few datasources for a bit too. The only part that I was getting stuck on was that for some reason if you don't set a particular (seemingly) unrealized monaco setting `scrollBeyondLastLine` to false then you end up in some sort of infinite resize loop where the code editor wants to expand forever. It sort of makes sense on reflection, but took me a while to sort out. It seems fine now, but it makes me a bit nervous to add this to grafana/ui as I'm not as confident as I'd like to be that this is a stable solution.

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/completion/statementPosition.ts:
```diff
@@ -0,0 +1,35 @@
+          lastComma.range.startColumn > currentFunction.range.startColumn &&
+          lastComma.range.startLineNumber >= currentFunction.range.startLineNumber;
+
+        if (lastCommaIsAfterSearch) {
```
Would be great if a unit test could be added here

sunker commented on public/app/plugins/datasource/cloudwatch/monarch/types.ts:
```diff
@@ -0,0 +1,99 @@
+  AfterOrderByDirection,
+  // metric math
+  PredefinedFunction,
+  SecondArgAfterSearchFunc,
```
Maybe these should be called `SearchFuncSecondArg` instead?

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/language.ts:
```diff
@@ -0,0 +1,156 @@
+  '||',
+];
+
+export const METRIC_MATH_PERIODS = [10, 60, 300, 900, 3500, 21600, 86400];
```
3500 it not a valid period. Use 3000 instead

sunker commented on public/app/plugins/datasource/cloudwatch/metric-math/completion/CompletionItemProvider.ts:
```diff
@@ -0,0 +1,116 @@
+          );
+
+        case SuggestionKind.Period:
+          METRIC_MATH_PERIODS.map((s) => addSuggestion(s.toString()));
```
It's unbelievable that monaco is not capable of sorting numeric values in order. :| Maybe you can find some way of working around that though. Something like this could work `[{insertText: 60, sortText: 'a'}, ...]`

![Screenshot from 2022-01-26 08-19-57](https://user-images.githubusercontent.com/2388950/151135119-a0df7bd7-ed71-4b56-9941-1a9609d6765b.png)

sunker reviewed (commented):
Just a few minor comments

Re the autosizing. Yeah I don't think we want to add stuff to the CodeEditor, so adding this to the data source sounds good to me. @gabor since you've done this before, do you want to take a look at the implementation? https://github.com/grafana/grafana/pull/43985/commits/5e5a22720c64254e4b20d6b37a00745d4d96de89

gabor commented:
> @gabor since you've done this before, do you want to take a look at the implementation?

it looks fine :+1:

sunker reviewed (approved):
> the auto-resize changes are fine +1

Thanks Gabor! 

I'm happy to merge as soon as my last comments are addressed. Again, amazing work on this @sarahzinger!

sarahzinger committed:
More CR fixes:
- fix missing break
- add unit tests for statementPosition
- fix broken time period
- sort time periods

sarahzinger committed:
Bug fix
